### PR TITLE
N:M Relation Editor

### PR DESCRIPTION
**Pull request by m-kuhn**
This adds the possibility to manage data on a normalized relational
database in N:M relations. On the relation editor in a form, the tools
to add, delete, link and unlink work (also) on the linking table if a
relation is visualized as N:M relation.

Configuration is done through the fields tab where on the relation a
second relation can be chosen (if there is a suitable relation in terms
of a second relation on the linking table).

nmrel
# Limitations

QGIS is not a database management system.

It is based on assumptions about the underlying database system. In
particular it expects
- A `ON DELETE CASCADE` or similar measure on the second relation. I.e. the linking entries are not deleted on "delete", only on "unlink".
- Does not take care of setting the primary key when adding features.
  Either users need to be instructed to set them manually or - if it's a
  database derived value - the layers need to be in transaction mode
  (currently only activatable through the API)

### CODE REVIEW

**Comment by mdouchin** [comment]
amazing !

**Comment by 3nids** [comment]
I can't compile (qt5) getting this error: `/src/core/qgstransaction.h:48:43: error: expected class name
class CORE_EXPORT QgsTransaction : public QObjec`
will try with qt4

**Comment by 3nids** [comment]
I can't link an existing feature.
A joining feature is created but is incomplete: it has the fk to the feature I was linking from but not to the feature I was linking to.

Same scenario with creating new feature (to link to).

Besides, with the uniefied-edit plugin, not all the layers seem to be turned in the transaction. weird.

**Comment by 3nids** [comment]
Also, the feature I was trying to link from was marked as edited, while it shouldn't

**Comment by m-kuhn** [comment]
When linking features, make sure that they are selected with the selection icon at the left of the list (yellow) and not as edited (blue). Note to self: should disable the buttons if no feature is selected. Or think of a better workflow.

For the unified edit plugin: are some connection strings different (e.g. sslmode differs, some layers use services and other db names)?

**Comment by 3nids** [comment]
Here are the troubles:

Linking without selecting leads to half empty entries in the joining table

Linking by selecting correctly fill the joining table, althgouh nothing is displayed in the widget from either just after linking or after reopening the widget

**Comment by m-kuhn** [comment]
Point number 2 is taken care of

**Comment by m-kuhn** [comment]
Point number 1 as well

**Comment by 3nids** [comment]
ready for testing again?

**Comment by m-kuhn** [comment]
yes

**Comment by m-kuhn** [comment]
Add feature still seems to fail on travis, the same tests passed for me. Not sure if it is related to pg versions but it's the only difference I see right now.

**Comment by 3nids** [comment]
now linking an existing feature is working.
I don't know why but the linked layer is marked as having editing features while I didn't do anything in this layer (hopefully qgis either)

If I link the same feature twice it is not listed in the relation editor widget although there has been a second row in the joining table. Reopening the linking feature form still show only one linked feature. This is an edge we might don't want to take care of as the db might not allow this. Although, I wonder if the feature should not be listed twice in such case.

adding also works properly.

For me it would be nice to merge, so we can test a bit more broadly. ok?

**Comment by medspx** [comment]
Hello,

first of all, thank you for trying to implement such a feature, it will make QGIS far better than other GIS software.

I've tested a little bit and I've got some segfaults but it is perhaps some configuration problems. Just to be sure: if you want to apply n,m relations, you have to use at least 3 tables and defines at least 2 relations ( towards the central table ) ?

For the moment it just displays a child form into another child form (or I have missed something). That's a good start but on the user experience side, it is a bit complicated and not really intuitive (too much forms). I have worked on a use case with n,m relations for a QGIS project. The users wanted to make a link between one geographic object and a multitude of values of another reference table. 

I've made [an article about my Python workaround and the UI choices I've made](http://medspx.fr/blog/Qgis/better_qgis_forms_part_three/). I think this is one important use case to implement (well perhaps not at the very beginning of n,m relations support) because on my project users are really happy about this feature: they have easy-to-use form controls that completely erase the database complex model.

In any case, your work is a good work and thank you for it !

**Comment by m-kuhn** [comment]
@3nids I guess that the layer appears to be edited is caused by an edit widget that erroneously signals changes of the value on initialization. So it's not related to this PR but would also need to be investigated (which widget, which field type, which value).

I haven't taken care of multiple links between the same rows. On the linking layer I normally have a compound primary key consisting of the two foreign keys, so the DB will effectively disallow this. I would say this is postponed to version 2.0.

Will merge as soon as I get the tests to pass.

@medspx Can you give more information concerning the segfaults? Segfaults are bad (you lose everything you didn't save), and if configuration is currently triggering it, it's not the config that's bad but the code not being able to deal with it.

Yes, n:m relations are expected to be relationally mapped like you wrote.

You will only get an embedded form in an embedded form if you configure the widgets on both ends of the relation. We may choose to avoid that by hiding the inner one based on the context.

File: tests/src/python/test_qgsrelationeditwidget.py [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+
+from utilities import (unitTestDataPath,
+                       getQgisTestApp,
+                       TestCase,
+                       unittest
+                       )
+QGISAPP, CANVAS, IFACE, PARENT = getQgisTestApp()
+
+
+class TestQgsTextEditWidget(TestCase):
```
Comment (m-kuhn) [inline tests/src/python/test_qgsrelationeditwidget.py]: Note to self: fixme

**Comment by m-kuhn** [comment]
Merged in be01b7eec5ec08069b731af7aca62d2b03bfbc40

Some tests that pass locally failed on travis previously. However due to the state travis is in currently it's impossible to debug this problem and I have to postpone this to when travis is back on track.

**Pull request closed without merging**